### PR TITLE
*WIP: Optimize `Selects`

### LINKED ISSUE #883: Optimize selects
E.g.: the expression
```
true ? 1 : 2;
```
can be changed to just `1`.

So basically, for every select expression, const evaluate the condition, and if the result is `true`, then just replace the select expression with the first branch. Otherwise, replace it with the second branch.

### PR DESCRIPTION
**Pull request by lethalgem**
Closes #883

### CODE REVIEW

**Comment by lethalgem** [comment]
I can't remember what was done in the past for pintc/src/asm_gen/tests.rs

Do we modify the test to account for the optimization pass? Or should the test input be modified to prevent optimizing out anything?

We have a flag for the tests that specifies that we don't skip the optimization phase, so I know we specifically want the phase to run.

File: pintc/tests/optimizations/dead_select_elimination.pnt [hunk 1]
```diff
@@ -0,0 +1,92 @@
+predicate test {
+    // select exprs that should be optimized out
+    var a = true ? 1 : 2;
+    var b = false ? 1 : 2;
+    var c = (1 + 2 == 3) ? 1 : 2;
+    var d = (1 < 2) ? 1 : 2;
+
+    // should this be removed? The evaluator can't handle this right now
+    var e = 2 < 1;
+    var f = e ? 1 : 2;
```
Comment (lethalgem) [inline pintc/tests/optimizations/dead_select_elimination.pnt:10]: Comment covers it here. We faced a similar issue with #834 where we currently don't remove `constraints` that have `vars` in the keys for the `constraint`.

It's currently just a limitation of the `evaluator` so maybe that should be opened as a new issue? Or maybe we don't want to remove things like this and `evaluator` is working as intended?

Comment (mohammadfawaz) [reply pintc/tests/optimizations/dead_select_elimination.pnt]: Yeah that's a more "advanced" evaluation.. let's not worry about it just yet, though it wouldn't be too difficult to do.

**Comment by mohammadfawaz** [comment]
> Do we modify the test to account for the optimization pass? 

Yes. We should make sure to change the tests to not trigger dead selection elimination, while keeping them as simple as possible (we don't want that file to blow up!)

File: pintc/src/warning.rs [hunk 2]
```diff
@@ -66,10 +66,19 @@ impl ReportableWarning for Warning {
...
             }
 
+
+
+
+
+
+
+
+
```
Comment (mohammadfawaz) [inline pintc/src/warning.rs:81]:
```suggestion
```

**Review by mohammadfawaz — COMMENTED** [review]
This is pretty good!

File: pintc/src/asm_gen/tests.rs [hunk 3]
```diff
@@ -179,10 +179,7 @@ fn select() {
                   Stack(Push(0))
                   Stack(Push(1))
                   Access(DecisionVarRange)
-                  Stack(Push(69))
```
Comment (mohammadfawaz) [inline pintc/src/asm_gen/tests.rs:182]: A reminder to change those tests to avoid selects from getting optimized away.

Comment (lethalgem) [reply pintc/src/asm_gen/tests.rs:182]: 🤦, fixed and didn't push

**Review by mohammadfawaz — APPROVED** [review]
Awesome :shipit:

**Pull request merged**